src/PoissonSolvers/LaplaceHelpers.h

```diff
@@ -0,0 +1,389 @@
+                : u_m(u)
+                , hvector_m(hvector)
+                , nghosts_m(nghosts)
+                , ldom_m(ldom)
+                , domain_m(domain) {}
+
+            /*
+             * n-dimensional upper+lower triangular Laplacian
+             */
+            template <typename... Idx>
+            KOKKOS_INLINE_FUNCTION auto operator()(const Idx... args) const {
+                using index_type = std::tuple_element_t<0, std::tuple<Idx...>>;
+                using T          = typename E::Mesh_t::value_type;
+                T res            = 0;
+                for (unsigned d = 0; d < dim; d++) {
+                    index_type coords[dim] = {args...};
+                    coords[d] -= 1;
+                    auto&& left = apply(u_m, coords);
+                    coords[d] += 2;
+                    auto&& right = apply(u_m, coords);
+                    res += hvector_m[d] * (left + right);
+                }
+                return res;
+            }
+
+        private:
+            using Mesh_t      = typename E::Mesh_t;
+            using Layout_t    = typename E::Layout_t;
+            using vector_type = typename Mesh_t::vector_type;
+            using domain_type = typename Layout_t::NDIndex_t;
+            const E u_m;
+            const vector_type hvector_m;
+            const unsigned nghosts_m;
+            const domain_type ldom_m;
+            const domain_type domain_m;
+        };
+    }  // namespace detail
+
+    /*!
+     * User interface of poisson
+     * @param u field
+     */
+    template <typename Field>
+    detail::meta_poisson<Field> poisson(Field& u) {
+        constexpr unsigned Dim = Field::dim;
+
+        u.fillHalo();
+        BConds<Field, Dim>& bcField = u.getFieldBC();
+        bcField.apply(u);
+
+        return detail::meta_poisson<Field>(u);
+    }
+
+    /*!
+     * User interface of lower triangular Laplacian
+     * @param u field
+     */
+    template <typename Field>
+    detail::meta_lower_laplace<Field> lower_laplace(Field& u) {
+        constexpr unsigned Dim = Field::dim;
+
+        u.fillHalo();
+        BConds<Field, Dim>& bcField = u.getFieldBC();
+        bcField.apply(u);
+
+        using mesh_type = typename Field::Mesh_t;
+        mesh_type& mesh = u.get_mesh();
+        typename mesh_type::vector_type hvector(0);
+        for (unsigned d = 0; d < Dim; d++) {
+            hvector[d] = 1.0 / std::pow(mesh.getMeshSpacing(d), 2);
+        }
+        const auto& layout = u.getLayout();
+        unsigned nghosts   = u.getNghost();
+        const auto& ldom   = layout.getLocalNDIndex();
+        const auto& domain = layout.getDomain();
+        return detail::meta_lower_laplace<Field>(u, hvector, nghosts, ldom, domain);
+    }
+
+    /*!
+     * User interface of lower triangular Laplacian without exchange of halo cells
+     * @param u field
+     */
+    template <typename Field>
+    detail::meta_lower_laplace<Field> lower_laplace_no_comm(Field& u) {
```

Comment:
The body of `lower_laplace_no_comm` is the same as `lower_laplace`, except for lines 230-232. This code duplication could be reduced e.g. by `lower_laplace` just calling `lower_laplace_no_comm(u)` after the fillHallo and BC apply steps (lines 230-232).

Reply:
I rewrote the code as you suggested.